doc/pandas-2.0/source/internal-architecture.rst

```diff
@@ -0,0 +1,714 @@
+            subarr = Categorical([value] * len(index))
+        else:
+            if not isinstance(dtype, (np.dtype, type(np.dtype))):
+                dtype = dtype.dtype
+            subarr = np.empty(len(index), dtype=dtype)
+            subarr.fill(value)
+
+or
+
+.. code-block:: python
+
+   if is_categorical_dtype(dtype):
+       upcast_cls = 'category'
+   elif is_datetimetz(dtype):
+       upcast_cls = 'datetimetz'
+   elif issubclass(dtype.type, np.bool_):
+       upcast_cls = 'bool'
+   elif issubclass(dtype.type, np.object_):
+       upcast_cls = 'object'
+   elif is_datetime64_dtype(dtype):
+       upcast_cls = 'datetime'
+   elif is_timedelta64_dtype(dtype):
+       upcast_cls = 'timedelta'
+   else:
+       upcast_cls = 'float'
+
+I've cherry-picked one of a number of places where this type of datatype-based
+branching happens.
+
+The primary reason for this complexity is that pandas is using both NumPy's
+dtype objects (which describe *physical storage*) as well as its own custom
+data type objects as a proxy for pandas's *semantic logical types*.
+
+Let's step back for a second and come up with clear language to steer the
+discussion.
+
+Some definitions
+~~~~~~~~~~~~~~~~
+
+Here is my attempt at definitions of some of the key terms:
+
+* **Metadata**: data that describes other data (such as its in-memory layout)
+
+* **Semantics**: The meaning / abstract interpretation of something. We often
+  discuss the semantics (meaning) of computer programs (i.e. what they do,
+  fundamentally) without touching upon low level details like machine
+  representation, programming languages, compilers, operating systems, etc.
+
+* **Physical data (or storage) types**: these are metadata objects which
+  provide a description of the precise structure of a piece of data in memory.
+
+  * In NumPy, the ``numpy.dtype`` object (aka ``PyArray_Descr`` in the C API)
+    is metadata describing a single cell / value in an array. Combined with the
+    ``shape`` and ``strides`` attributes of the ``ndarray`` object, you have
+    enough information to perform O(1) random access on any cell in an
+    ``ndarray`` and to assign these values to a C type (or, in the case, of
+    structured dtypes, assign to a packed C struct).
+
+  * This may or may not include a physical representation of NULL or missing
+    data (for example: nullable float64 might be a physical type indicating a
+    normal float64 array along with a bitmap of null/not-null indicators).
+
+* **Logical data type**: metadata which describes the semantic content of a
+  single value in an array or other collection of values. Depending on the
+  logical type, it may map 1-to-1 to a physical type or not at all. Here are
+  some examples:
+
+  * The ``double`` or ``float64`` type may be viewed both as a logical type as
+    well as a physical type (a 1-to-1 correspondence).
+
+  * pandas's ``category`` dtype contains its own auxiliary array of category
```

Comment:
Let me rephrase that -- what I meant by "physical memory representation" was describing the "shape" or "physical structure" of an array cell. The actual memory buffers in an array would need to be defined in an array library implementation like libdynd